Image RTL

Hixie opened the pull request.

Hixie committed:
Image RTL

Hixie commented:
cc @abarth

Hixie commented:
Fixes https://github.com/flutter/flutter/issues/11391 https://github.com/flutter/flutter/issues/11589 https://github.com/flutter/flutter/issues/11999 https://github.com/flutter/flutter/issues/12001

abarth commented on packages/flutter/lib/src/painting/box_decoration.dart:
```diff
@@ -321,9 +321,26 @@ class _BoxDecorationPainter extends BoxPainter {
+          // false is the default
+          break;
+      }
+    }
```
I'd factor this out into a `textDirectionIsReversed` analogous to `axisDirectionIsReversed`.  Then this becomes:

```
bool flipHorizontally = !backgroundImage.matchTextDirection || textDirectionIsReversed(configuration.textDirection);
```

abarth replied:
I take it back.  We don't want other clients to call `textDirectionIsReversed`.

abarth commented on packages/flutter/lib/src/rendering/image.dart:
```diff
@@ -182,6 +193,22 @@ class RenderImage extends RenderBox {
 
+  /// Whether the image should be flipped horizontally.
+  ///
+  /// This is occasionally used with images in right-to-left environments, for
```
I'd put something in here with `[TextDirection]` so that folks reading the docs can get back to the "main article".

Hixie replied:
done here and in RawImage

abarth commented on packages/flutter/lib/src/widgets/fade_in_image.dart:
```diff
@@ -442,6 +475,20 @@ class _FadeInImageState extends State<FadeInImage> with TickerProviderStateMixin
+        return false;
+    }
+    return null;
+  }
```
It's a bummer to have this code three times.  Maybe pull it out into a static function on `RawImage`?  That's probably better than a free `textDirectionIsReversed` function because folks will be less tempted to use it for RTL generally.

Hixie replied:
It's a lot of words, but the actual code is trivial here. It's basically a very verbose ternary operator. Would you feel the same way if I wrote it as this?:

```dart
   assert(!widget.matchTextDirection || debugCheckHasDirectionality(context));
   return widget.matchTextDirection && Directionality.of(context) == TextDirection.rtl;
```

It's basically identical to this logic, I'm just avoiding writing that because it's an enum and I don't want to set the pattern of using direct comparisons on enums (as opposed to using switch). (Though it's unclear to me whether that's worth it for this case; we're not likely to add another text direction.)

abarth replied:
The reason not to use direct comparisons is that the enum really has three values (RTL, LTR, and null).  Even if you assert, in production you could have the null value.  When that happens, you want an exception of some sort instead of silently assuming LTR (or RTL).

Hixie replied:
You don't necessarily get an exception in release mode even with the switch. For example, when I had the switch in the ImageDecoration code, it did the same thing for null and for LTR, which is to say, nothing, leaving it at the default (false).

abarth replied:
Yeah, I thought about that when reading the code, but it was below the threshold of things that I usually comment on.  :)

Hixie replied:
Well in any case, I don't think we should optimize for getting exceptions when things are broken in release mode. If your code would assert in debug builds, then behaviour in release builds is undefined, IMHO.

abarth commented on packages/flutter/lib/src/widgets/fade_in_image.dart:
```diff
@@ -454,8 +501,9 @@ class _FadeInImageState extends State<FadeInImage> with TickerProviderStateMixin
       colorBlendMode: BlendMode.modulate,
       fit: widget.fit,
-      alignment: widget.alignment,
+      alignment: widget.alignment is FractionalOffset ? widget.alignment : widget.alignment.resolve(Directionality.of(context)),
```
This is less efficient than doing the resolve on the `RenderImage` because if you're using a `FractionalOffsetDirectional`, you'll re-resolve every build where as the `RenderImage` pattern only re-resolves when either the `alignment` or the `textDirection` change.

Hixie replied:
if I'm going to send a textDirection down there anyway, I guess I can move all the flipHorizontally logic there too...

Hixie replied:
(that'll bring it from 3 sites to 2)

abarth commented on packages/flutter/test/widgets/image_rtl_test.dart:
```diff
@@ -0,0 +1,544 @@
+      ..clipRect(rect: new Rect.fromLTRB(0.0, 0.0, 100.0, 50.0))
+      ..translate(x: 50.0, y: 0.0)
+      ..scale(x: -1.0, y: 1.0)
+      ..translate(x: -50.0, y: 0.0)
```
Oh, we already have the transform ops.

abarth commented:
Looks pretty good, but I'd move the `alignment` resolve to `RenderImage` to follow the pattern we use elsewhere (and for efficiency).

abarth commented:
I'm headed to sleep, but LGTM for whatever you decide to do about the code duplication.  Getting it down to two might be enough.  The general rule of thumb I like is that three copies is when its worth thinking hard about how to share.

Hixie commented:
I found an interesting bug when moving stuff to RenderImage, which affected all the other render objects that took textDirection and used it before layout (so Stack, Padding, Align, etc): if you change the textDirection to null in the same frame as you make the textDirection unnecessary, you throw (or if you don't, then you instead throw on the frame where you change textDirection to non-null and make it necessary).

I've fixed that, added some tests for it too (for Stack, Padding, and Image; I didn't test the other widgets), but the fix is ugly: the updateRenderObject methods have to actually verify whether the textDirection is non-null, and if so, set it _before_ the other properties, otherwise, it has to be set last.

PTAL.

abarth commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -1129,9 +1129,12 @@ class Padding extends SingleChildRenderObjectWidget {
   void updateRenderObject(BuildContext context, RenderPadding renderObject) {
+    final TextDirection textDirection = Directionality.of(context);
+    if (textDirection != null)
+      renderObject.textDirection = textDirection; // in case it was null before, we must set it before anything that depends on it
```
Hum...  This pattern is pretty ugly.  Here are two alternatives:

1. Delay resolving the padding/alignment until the resolve value is actually needed.  This will give both writes a chance to land before we resolve (avoiding the problem of resolving in an inconsistent state).

2. Add a method to set the padding/alignment and textDirection at the same time.  That will like the client perform an atomic write of both properties an avoid the inconsistent state entirely.

Hixie replied:
Number 1 is my preferred approach, but it means being more aggressive about marking things dirty. I'll try that.

Hixie commented:
Ok I apply #1 everywhere. PTAL.

abarth commented on packages/flutter/lib/src/rendering/image.dart:
```diff
@@ -28,22 +32,47 @@ class RenderImage extends RenderBox {
     _updateColorFilter();
   }
 
+  bool _needsResolution = true;
```
You might want to use the nullity of _resolvedAlignment for this bit.  That also stops code from using stale values because they'll be nulled out.

Hixie replied:
Ah, yes. That's the solution I was looking for. Thanks.

abarth commented on packages/flutter/lib/src/rendering/shifted_box.dart:
```diff
@@ -98,20 +98,22 @@ class RenderPadding extends RenderShiftedBox {
 
-  // The resolved absolute insets.
+  bool _needsResolution = true;
   EdgeInsets _resolvedPadding;
```
Same thought here about using the nullity of _resolvedPadding to represent that the value needs to be computed.

abarth commented:
![LGTM](http://i3.cpcache.com/product/1351564863/lgtm_bumper_stickers.jpg?width=200&height=200)

Hixie commented:
Switched to using nullness instead of an explicit flag.

Hixie commented:
thanks for the review, will check in on green
